### PR TITLE
Implement Generate Hierarchical CV DB

### PR DESCRIPTION
**Pull request by itsfarseen**
Depends on #13

### CODE REVIEW

File: generate_hierarchical_cv_db/__main__.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+
+# Not used because need to figure out how to batch tensors of different sizes
+# def make_batches(xs, batch_size):
+#     for i in range(0, len(xs), batch_size):
+#         yield xs[i:i+batch_size]
+# 
+
+def main():
+    compressor_chkpt = "../rigel-data/hierarchical-compression-checkpoint-2024-10-30/attention_model.pt"
```
Comment (nalane) [inline generate_hierarchical_cv_db/__main__.py]: This should be an argument passed to `main`

File: generate_hierarchical_cv_db/__main__.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    compressor = Compressor(compressor_chkpt)
+    #batch_size = 1024
+
+    input_db_path = Path("../rigel-data/context-vectors-2024-10-30/")
+    output_db_path = Path("../rigel-data/context-vectors-compressed-2024-10-30")
```
Comment (nalane) [inline generate_hierarchical_cv_db/__main__.py]: These should also be arguments

File: generate_hierarchical_cv_db/__main__.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
+from cv_library.compressor import Compressor
```
Comment (nalane) [inline generate_hierarchical_cv_db/__main__.py]: I'd actuall prefer this not to be a module, but a runnable file `generate_hierarchical_cv_db.py` in the root

File: generate_hierarchical_cv_db/__main__.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        for section_name in db.get_section_names(article_title):
+            res.append((article_title, section_name))
+    return res
+
+
+# Not used because need to figure out how to batch tensors of different sizes
+# def make_batches(xs, batch_size):
+#     for i in range(0, len(xs), batch_size):
+#         yield xs[i:i+batch_size]
+#
```
Comment (nalane) [inline generate_hierarchical_cv_db/__main__.py]:
```suggestion
```

File: generate_hierarchical_cv_db/__main__.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        shutil.rmtree(output_db_path)
+    except FileNotFoundError:
+        pass
+    output_db_path.mkdir()
+
+
+    input_db = ContextVectorDB(input_db_path)
+    output_db_config = cvhs.DBConfig(
+            vec_max_size = 4096,
+            vec_min_size = 8,
```
Comment (nalane) [inline generate_hierarchical_cv_db/__main__.py]: This value should be, by default, `cv_library.hierarchical_compression.Attention.NUM_HEADS`

File: generate_hierarchical_cv_db/__main__.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            compression_factor = 8,
```
Comment (nalane) [inline generate_hierarchical_cv_db/__main__.py]: This should also be a parameter

File: generate_hierarchical_cv_db/__main__.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            compression_dimension = 1, # cv.shape = (seq_size, 4096). We compress the one with 4096.
+            search_narrow_factor = 16,
```
Comment (nalane) [inline generate_hierarchical_cv_db/__main__.py]: This should also be a parameter

File: generate_hierarchical_cv_db/__main__.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    )
+    output_db = cvhs.Database(output_db_path, output_db_config, sequence_similarity)
+
+    print("Reading metadata ..")
+    headings = get_all_headings(input_db)
+    # headings_batched = list(make_batches(headings, batch_size))
```
Comment (nalane) [inline generate_hierarchical_cv_db/__main__.py]:
```suggestion
```

File: generate_hierarchical_cv_db/__main__.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    print("Processing ..")
+    # for batch in tqdm(headings_batched):
+    #     cvs = []
+    #     for article_title, section_name in batch:
+    #         cvs.append(input_db.get(article_title, section_name))
+    #     cvs = np.stack(cvs)
+    #     cvs = torch.tensor(cvs)
+    #     print("cvs", cvs.shape)
+    #     output_cvs = compressor.compress(cvs)
+    #     print("output_cvs", output_cvs.shape)
```
Comment (nalane) [inline generate_hierarchical_cv_db/__main__.py]:
```suggestion
```

File: generate_hierarchical_cv_db/__main__.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        cv = input_db.get(article_title, section_name)
+        cv = torch.tensor(cv).unsqueeze(dim=0)
+        output_cvs = [cv, *compressor.compress(cv)]
+        output_cvs_fixed = [cv.squeeze(dim=0).cpu().detach().numpy() for cv in output_cvs]
+        output_cvs_fixed.reverse()
+        output_db.insert(output_cvs_fixed)
+
+
+if __name__ == "__main__":
+    main()
```
Comment (nalane) [inline generate_hierarchical_cv_db/__main__.py]: You can use Fire to make it so that the parameters to this main function become command line arguments

**Review by nalane — CHANGES_REQUESTED** [review]

File: generate_hierarchical_cv_db/__main__.py (hunk 5 above)
Comment (itsfarseen) [reply generate_hierarchical_cv_db/__main__.py]: NUM_HEADS seem to be 32. I'm not sure how to use it here.

I've also changed the parameters to DBConfig. Now there's a single `level_sizes` parameter, which is specified in the `justfile` as:

```
--level_sizes 16,64,256,1024,4096
```

Does this work?

Comment (itsfarseen) [reply generate_hierarchical_cv_db/__main__.py]: I got these values by inspecting the shape of the compressor's output.

**Review by nalane — APPROVED** [review]

**Comment by nalane** [comment]
Looks great!

**Pull request merged**